myia/debug/finite_diff.py

```diff
@@ -0,0 +1,270 @@
+"""Estimate gradients with finite differences."""
+
+
+from typing import Callable, Dict, List, Any
+import numpy
+import itertools
+
+from ..utils import smap, overload
+
+
+@smap.variant
+def _zeros_like(self, x: object):
+    return 0
+
+
+class NoTestGrad:
+    """Class to signify that a value's gradient shouldn't be tested.
+
+    I don't fully remember how this is supposed to work.
+    """
+
+    def __init__(self, value):
+        """Initialize NoTestGrad."""
+        self.value = value
```

Comment:
Support for NoTestGrad seems to make this ridiculously more complex than it should be.